### ISSUE TITLE
Add DynamicallyAccessedMemberTypes sugar members over existing members

### ISSUE THREAD

**Issue by roji** [dotnet/linker#2488]
It's fairly typical to want to preserve all properties, or all members on a type; this currently requires OR'ing together multiple enum values from DynamicallyAccessedMemberTypes:

```c#
[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicProperties)]
```

As these annotations propagate around the codebase, things start to get verbose. This could be made nicer by introducing enum members which OR other members:

```c#
[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.AllProperties)]
```

This could be done for all public/non-public pairs, as well as maybe AllMembers for both properties and fields (there may be some other opportunities).

The trimming analyzer could even have a code fix to replace the verbose form with the shorter one.

**Comment by MichalStrehovsky** [comment]
`DynamicallyAccessedMemberTypes` is modeled after `BindingFlags`, because that's where this eventually leads to. AFAIK there's no better way to get both public and non-public properties than to specify the verbose `BindingFlags.Public | BindingFlags.NonPublic` so this is kind of par for the course. I kind of like that it makes people more careful about what they actually want...

**Comment by roji** [comment]
> DynamicallyAccessedMemberTypes is modeled after BindingFlags, because that's where this eventually leads to.

I understand that, but is consistency BindingFlags a good reason to not add enum values to DynamicallyAccessedMemberTypes that make it less painful to work with?

For one thing, BindingFlags are used only in the call site actually doing the reflection call (e.g. GetMethod). However, the trimming annotations are viral - any method calling a method with a DynamicallyAccessedMembers parameter must itself duplicate those values so as to bubble them up. This can introduce a *lot* of duplicated code into the codebase, and terseness does seem to be important.

For another example, at least at higher levels of the API, multiple individual DynamicallyAccessedMemberTypes sometimes get aggregated. A serializer/deserializer (such as EF Core) typically needs to access both properties and fields (which would get populated at runtime) - this alone already forces specifying 4 values (properties and fields, both public and non-public) in many APIs (which again, are just built over APIs that do the actual reflection).

Finally, there already do exist reflection APIs such as TypeInfo.GetDeclaredMethod, which are themselves annotated with e.g. PublicMethods | NonPublicMethods (I realize TypeInfo is obsolete, but...). So this once again forces users into a needless verbosity where the API being used doesn't even require BindingFlags at all.

> I kind of like that it makes people more careful about what they actually want...

That's a good point, though I think an enum value name such as AllProperties (with All) would at least make it clear that one is possibly including more properties than is needed, and that a more fine-grained option exists.

My concern here is that the amount of noise added to a codebase in order to make it trimming-friendly may deter some developers from doing so - it seems important for trimming-friendliness to be as minimally-invasive as it can be.

**Comment by MichalStrehovsky** [comment]
> For one thing, BindingFlags are used only in the call site actually doing the reflection call (e.g. GetMethod). However, the trimming annotations are viral - any method calling a method with a DynamicallyAccessedMembers parameter must itself duplicate those values so as to bubble them up. This can introduce a lot of duplicated code into the codebase, and terseness does seem to be important.

In framework code, we introduced `consts` with the appropriate value. Framework also does this with `BindingFlags` in some spots. Different use cases have different needs for annotations. E.g. properties and parameterless constructor is another common one, properties and fields is common too, as you write.

**Comment by roji** [comment]
> In framework code, we introduced consts with the appropriate value.

Yeah, I'm also going down this path - that seems to be an indication that at least some common values could be useful and make the annotation process easier. I also suspect many annotators would miss the possibility of using consts here.

Anyway, the idea is clear... If you don't this is a good idea, feel free to close this issue.

**Comment by vitek-karas** [comment]
My main concern is actually that it promotes private reflection - which I think should be considered evil.
For example, LOT of people don't realize that private fields/properties are not visible when doing reflection over derived types. So take this as an example:
```C#
class Base
{
    private string _baseField;
}

class Derived : Base
{
    private string _derivedField;
}

void Main()
{
    TestMethod(typeof(Derived)); // This will NOT preserve _baseField
}

void TestMethod([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicFields)] Type type)
{
    type.GetFields(); // Only returns _derivedField (even without trimming)
}
```

We have only one annotation which preserves even private members on base types. `DynamicallyAccessedMembers.All` does that.

If we were to introduce `DynamicallyAccessedMembers.AllProperties` then it should probably behave more like the `All` annotation, and preserve all properties on all base types (interfaces).

(Personally I don't like any serializer/deserializer which works on private fields/properties... it breaks encapsulation principles... - but that's just my rant :-))

I'm split on this idea - I understand the desire to keep things short... but in this case...

**Comment by roji** [comment]
Is the private fields/properties on derived types the only case/problem you see? I'd argue that this "pit of failure" is already there with `NonPublicFields` (and properties), since the expectations is for that to presserve the derived ones as well. And is, it may be OK for `AllProperties` to be defined simply as `PrivateProperties | PublicProperties`, rather than assuming it has to behave more like `All` (in that it needs to preserve the derived).

Aside from that, the derived issue seems like a bit of an edge case, I wouldn't necessarily shoot down the general improvement just because of that...
